Add torch._C.Tag.needs_contiguous_strides

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152859

this forces inductor to force the inputs to be contiguous.

Test Plan:
- new test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

zou3519 committed:
Add torch._C.Tag.needs_contiguous_strides

this forces inductor to force the inputs to be contiguous.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch._C.Tag.needs_contiguous_strides"


this forces inductor to force the inputs to be contiguous.

Test Plan:
- new test

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -164,6 +164,8 @@ def maybe_layout_constraints(fn: Callable[..., Any]) -> Optional[Callable[..., A
     if tag == torch._C.Tag.needs_exact_strides:
         return constrain_to_fake_tensors
+    if tag == torch._C.Tag.needs_contiguous_strides:
+        return require_contiguous
```
Can we construct the strides manually ? I had a [pr](https://github.com/pytorch/pytorch/pull/148235) to fix this but it was blocked by mkldnn tests. @CaoE has a pr to update https://github.com/pytorch/pytorch/pull/148424.  cc @CaoE @leslie-fang-intel we really need to land that.

eellison reviewed (commented):
Needs one small change until we land other pr

zou3519 committed:
Update on "Add torch._C.Tag.needs_contiguous_strides"


this forces inductor to force the inputs to be contiguous.

Test Plan:
- new test

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -i
